In [780]:
from astropy.io import fits
import numpy as np
from scipy.optimize import curve_fit
import math

In [814]:
s = 'hi'
try:
    h = s.index('h')
    b = s.index('b')
    s = s[h:b]
except:
    print('oops')
print(s)

oops
hi


In [801]:
file = fits.open('../data/250611_03:56:34/1555+111_full.fits')

In [782]:
data = file[0].data[0]

In [783]:
data

array([[-0.00097676,  0.00081531,  0.0024485 , ...,  0.00310269,
         0.00357472,  0.0038019 ],
       [ 0.00050492,  0.00174901,  0.00297046, ...,  0.00181118,
         0.00253195,  0.00331035],
       [ 0.00139358,  0.00180249,  0.00227913, ..., -0.00087274,
        -0.00017827,  0.00090418],
       ...,
       [ 0.00517534,  0.00605375,  0.00648655, ..., -0.00436939,
        -0.0046905 , -0.00504718],
       [ 0.00252144,  0.00371739,  0.00465888, ..., -0.00455041,
        -0.00441014, -0.00434385],
       [-0.00199351, -0.00084242,  0.00030834, ..., -0.00385592,
        -0.00336875, -0.0029433 ]], dtype='>f4')

In [784]:
peak = data.max()

In [785]:
peak_pix = np.where(data == peak)

In [786]:
peak_pix

(array([214]), array([214]))

In [787]:
peak_x = peak_pix[0][0]
peak_y = peak_pix[1][0]

In [788]:
max_x = len(data)
max_y = len(data[0])

In [789]:
if (peak_x - 2) >= 0 and (peak_x + 2) <= max_x and (peak_y - 2) >= 0 and (peak_y + 2) <=max_y:
    neg2_2 = data[peak_x - 2][peak_y + 2]
    neg2_1 = data[peak_x - 2][peak_y + 1]
    neg2_0 = data[peak_x - 2][peak_y]
    neg2_neg1 = data[peak_x - 2][peak_y - 1]
    neg2_neg2 = data[peak_x - 2][peak_y - 2]
    neg1_2 = data[peak_x - 1][peak_y + 2]
    neg1_1 = data[peak_x - 1][peak_y + 1]
    neg1_0 = data[peak_x - 1][peak_y]
    neg1_neg1 = data[peak_x - 1][peak_y - 1]
    neg1_neg2 = data[peak_x - 1][peak_y - 2]
    zero_2 = data[peak_x][peak_y + 2]
    zero_1 = data[peak_x][peak_y + 1]
    zero_neg1 = data[peak_x][peak_y - 1]
    zero_neg2 = data[peak_x][peak_y - 2]
    pos1_2 = data[peak_x + 1][peak_y + 2]
    pos1_1 = data[peak_x + 1][peak_y + 1]
    pos1_0 = data[peak_x + 1][peak_y]
    pos1_neg1 = data[peak_x + 1][peak_y - 1]
    pos1_neg2 = data[peak_x + 1][peak_y - 2]
    pos2_2 = data[peak_x + 2][peak_y + 2]
    pos2_1 = data[peak_x + 2][peak_y + 1]
    pos2_0 = data[peak_x + 2][peak_y]
    pos2_neg1 = data[peak_x + 2][peak_y - 1]
    pos2_neg2 = data[peak_x + 2][peak_y - 2]

    '''left_top = data[peak_x - 1][peak_y + 1]
    left_middle = data[peak_x - 1][peak_y]
    left_bottom = data[peak_x - 1][peak_y - 1]
    middle_top = data[peak_x][peak_y + 1]
    middle_bottom = data[peak_x][peak_y - 1]
    right_top = data[peak_x + 1][peak_y + 1]
    right_middle = data[peak_x + 1][peak_y]
    right_bottom = data[peak_x + 1][peak_y - 1]'''

    #z_data = [left_top, left_middle, left_bottom, middle_top, peak, middle_bottom, right_top, right_middle, right_bottom]
    #x_data = [-1]*3 + [0]*3 + [1]*3
    #y_data = [1, 0, -1] * 3

    z_data = [neg2_2, neg2_1, neg2_0, neg2_neg1, neg2_neg2,\
              neg1_2, neg1_1, neg1_0, neg1_neg1, neg1_neg2,\
              zero_2, zero_1, peak, zero_neg1, zero_neg2,\
              pos1_2, pos1_1, pos1_0, pos1_neg1, pos1_neg2,\
              pos2_2, pos2_1, pos2_0, pos2_neg1, pos2_neg2]
    x_data = [-2]*5 + [-1]*5 + [0]*5 + [1]*5 + [2]*5
    y_data = [2, 1, 0, -1, -2]*5

In [790]:
def paraboloid(coord, a, b, c, d, e, f):
    x, y = coord
    return a*x**2 + b*y**2 + c*x*y + d*x + e*y + f

In [791]:
'''%%timeit
popt, pcov = curve_fit(paraboloid, (x_data, y_data), z_data)
a, b, c, d, e, f = popt

local_max = [float('-inf'), None, None] #flux value, RA, Dec

x_vals = np.linspace(-1, 1, 2001)
y_vals = np.linspace(-1, 1, 2001)
for x in x_vals:
    for y in y_vals:
        z = paraboloid((x,y), a, b, c, d, e, f)
        if z > local_max[0]:
            local_max = [float(z), float(x), float(y)]'''

"%%timeit\npopt, pcov = curve_fit(paraboloid, (x_data, y_data), z_data)\na, b, c, d, e, f = popt\n\nlocal_max = [float('-inf'), None, None] #flux value, RA, Dec\n\nx_vals = np.linspace(-1, 1, 2001)\ny_vals = np.linspace(-1, 1, 2001)\nfor x in x_vals:\n    for y in y_vals:\n        z = paraboloid((x,y), a, b, c, d, e, f)\n        if z > local_max[0]:\n            local_max = [float(z), float(x), float(y)]"

In [792]:
def gaussian_theta(coord, amp, sigma, theta, mu_x, mu_y):
    x, y = coord
    return amp * np.exp(-(((x-mu_x)*math.cos(theta)+(y-mu_y)*math.sin(theta))**2+(-(x-mu_x)*math.sin(theta)+(y-mu_y)*math.cos(theta))**2)/(2*sigma**2))

In [793]:
def gaussian(coord, amp, sigma, mu_x, mu_y):
    x, y = coord
    return amp * np.exp(-((x-mu_x)**2+(y-mu_y)**2)/(2*sigma**2))

In [794]:
def rotated_gaussian(coord, amp, sigma, mu_x, mu_y, a, b, c, d, e):
    x, y = coord
    return -amp * np.exp(-(a*(x-mu_x)**2+b*(y-mu_y)**2+c*(x-mu_x)*(y-mu_y)+d*(x-mu_x)+e*(y-mu_y))/(2*sigma**2))

In [ ]:
#%%timeit
popt, pcov = curve_fit(gaussian_theta, (x_data, y_data), z_data)
amp, sigma, theta, mu_x, mu_y = popt

local_max = [float(gaussian_theta((mu_x, mu_y), amp, sigma, theta, mu_x, mu_y)), float(mu_x), float(mu_y)]

In [796]:
popt

array([ 1.71732832e-01,  2.31195090e+00,  5.66136912e+10, -4.13900264e-01,
        1.08018079e-01])

In [797]:
local_max

[0.17173283170048942,
 np.float64(-0.4139002641544724),
 np.float64(0.10801807882243354)]

In [798]:
local_max[1] += peak_x
local_max[2] += peak_y

In [799]:
local_max = [float(ele) for ele in local_max]

In [800]:
local_max

[0.17173283170048942, 213.58609973584552, 214.10801807882243]

3x3:
[0.17188927636389958, 213.556, 214.111] gaussian, 1.26 ms ± 151 μs
[0.1698465144295369, 213.749, 214.197] paraboloid, 4.66 s ± 27 ms
[0.1718892811816014, 213.556, 214.111] rotated gaussian with theta, 1.91 ms ± 43.8 μs
[0.17076217940215302, 213.737, 214.206] rotated gaussian with paraboloid, 6.29 ms ± 100 μs

5x5:
[0.17173282801699105, 213.586, 214.108] gaussian, 874 μs ± 7.57 μs
[0.16409301582673863, 213.758, 214.186] paraboloid 4.59 s ± 9.45 ms
[0.17173283153544183, 213.586, 214.108] rotated gaussian with theta, 1.72 ms ± 165 μs
[0.17140147598132904, 213.722, 214.209] rotated gaussian with paraboloid, 3.9 ms ± 58.1 μs